internal/command/deploy/plan.go

```diff
@@ -172,39 +200,90 @@ func (md *machineDeployment) updateMachinesWRecovery(ctx context.Context, oldApp
 		}
 
 		defer func() {
-			err := md.releaseLeases(ctx, machineTuples, sl)
+			err := md.releaseLeases(ctx, machineTuples, machineLogger)
 			if err != nil {
 				fmt.Fprintln(md.io.ErrOut, "Failed to release leases:", err)
 				span.RecordError(err)
 			}
 		}()
 	}
 
-	statusLines := map[string]statuslogger.StatusLine{}
-	for idx, machPair := range machineTuples {
-		if machPair.oldMachine != nil {
-			statusLines[machPair.oldMachine.ID] = sl.Line(idx)
-		} else if machPair.newMachine != nil {
-			statusLines[machPair.newMachine.ID] = sl.Line(idx)
-		}
-	}
-
 	pgroup := errgroup.Group{}
 	pgroup.SetLimit(rollingStrategyMaxConcurrentGroups)
 
 	// We want to update by process group
 	for _, machineTuples := range machPairByProcessGroup {
 		machineTuples := machineTuples
 		pgroup.Go(func() error {
-			err := md.updateProcessGroup(ctx, machineTuples, statusLines, poolSize)
-			if err != nil && strings.Contains(err.Error(), "lease currently held by") {
-				err := &unrecoverableError{err: err}
+
+			errChan := make(chan error, 2)
+			wg := sync.WaitGroup{}
+
+			warmMachines := lo.Filter(machineTuples, func(e machinePairing, i int) bool {
+				if e.oldMachine != nil && e.oldMachine.State == "started" {
+					return true
+				}
+				if e.newMachine != nil && e.newMachine.State == "started" {
+					return true
+				}
+				return false
+			})
+
+			coldMachines := lo.Filter(machineTuples, func(e machinePairing, i int) bool {
+				if e.oldMachine != nil && e.oldMachine.State != "started" {
+					return true
+				}
+				if e.newMachine != nil && e.newMachine.State != "started" {
+					return true
+				}
+				return false
+			})
+
```

Comment:
Perhaps another opportunity to use `errgroup`

Reply:
https://github.com/superfly/flyctl/pull/3817#discussion_r1707001486